include/aws/s3/private/s3_util.h

```diff
@@ -323,6 +323,15 @@ int aws_s3_crt_error_code_from_recoverable_server_error_code_string(struct aws_b
 AWS_S3_API
 void aws_s3_request_finish_up_metrics_synced(struct aws_s3_request *request, struct aws_s3_meta_request *meta_request);
 
+/* Check the response headers for checksum to verify, return a running checksum based on the algorithm found. If no
+ * checksum found from header, return null. */
+AWS_S3_API
+struct aws_s3_checksum *aws_s3_check_headers_for_checksum(
```

Comment:
seems like deep in the callstack stuff like `aws_sha1_default_new()` can fail (but we're not checking it). So you might need better indication of "error" vs "no need to do work". And we also probably should be checking this stuff for failure.